File: java/sodium/src/sodium/LazyCell.java
package sodium;

class LazyCell<A> extends Cell<A> {
    LazyCell(final Stream<A> event, final Lazy<A> lazyInitValue) {
        super(event, null);
        this.lazyInitValue = lazyInitValue;
    }

    @Override
    protected A sampleNoTrans()
    {
        if (value == null && lazyInitValue != null) {
            value = lazyInitValue.get();
            lazyInitValue = null;
        }
        return value;
    }
}

File: c#/src/Sodium/Sodium/LazyCell.cs
using System;

namespace Sodium
{
    public class LazyCell<T> : Cell<T>
    {
        internal Lazy<T> LazyInitialValue;

        internal LazyCell(Stream<T> stream, Lazy<T> lazyInitialValue)
            : base(stream, default(T)) => this.LazyInitialValue = lazyInitialValue;

        protected override void NotUsingInitialValue()
        {
            base.NotUsingInitialValue();

            this.LazyInitialValue = null;
        }

        internal override T SampleNoTransaction()
        {
            if (this.UsingInitialValue && this.LazyInitialValue != null)
            {
                this.ValueProperty = this.LazyInitialValue.Value;
                this.LazyInitialValue = null;
            }

            return this.ValueProperty;
        }
    }
}